File: euler36.py
# Найдите сумму всех чисел меньше миллиона, являющихся палиндромами
# по основаниям 10 и 2.

from datetime import datetime


def check_palindrom(number: str) -> bool:
    """
    Проверка заданного числа на палиндром.

    Args:
        number: Строковое представление числа.

    Returns:
        True если число является палиндромом, иначе False.
    """
    return number == number[::-1]


def euler36() -> None:
    """
    Решение задачи Эйлера №36.

    Найдите сумму всех чисел меньше миллиона, являющихся палиндромами
    по основаниям 10 и 2.
    """
    n = 1_000_000

    # 1 вариант
    start_time = datetime.now()
    sum_db_palindromes = 0
    for i in range(n):
        if check_palindrom(str(i)):
            if check_palindrom(str(bin(i)[2:])):
                sum_db_palindromes += i
    print(sum_db_palindromes)
    print(datetime.now() - start_time)

    # 2 вариант
    start_time = datetime.now()
    print(sum(
        x for x in range(n)
        if check_palindrom(str(x)) and check_palindrom(str(bin(x)[2:]))
    ))
    print(datetime.now() - start_time)


if __name__ == "__main__":
    euler36()

File: go/euler36.go
// euler36 находит сумму всех чисел меньше миллиона, являющихся
// палиндромами по основаниям 10 и 2.
//
// Решение задачи Эйлера №36.
package main

import (
	"fmt"
	"time"
)

func checkPalindrom(number string) bool {
	for i := 0; i < len(number)/2; i++ {
		if number[i] != number[len(number)-1-i] {
			return false
		}
	}
	return true
}

func toBinary(n int) string {
	if n == 0 {
		return "0"
	}
	result := ""
	for n > 0 {
		result = fmt.Sprintf("%d", n%2) + result
		n /= 2
	}
	return result
}

func euler36() {
	n := 1000000

	startTime := time.Now()
	sumDbPalindromes := 0
	for i := 0; i < n; i++ {
		if checkPalindrom(fmt.Sprintf("%d", i)) {
			if checkPalindrom(toBinary(i)) {
				sumDbPalindromes += i
			}
		}
	}
	fmt.Println(sumDbPalindromes)
	fmt.Println(time.Since(startTime))

	startTime = time.Now()
	sumDbPalindromes = 0
	for i := 0; i < n; i++ {
		if checkPalindrom(fmt.Sprintf("%d", i)) && checkPalindrom(toBinary(i)) {
			sumDbPalindromes += i
		}
	}
	fmt.Println(sumDbPalindromes)
	fmt.Println(time.Since(startTime))
}

func main() {
	euler36()
}
